core/src/main/scala/chisel3/Aggregate.scala

```diff
@@ -297,6 +297,11 @@ sealed class Vec[T <: Data] private[chisel3] (gen: => T, val length: Int) extend
     requireIsHardware(this, "vec")
     requireIsHardware(p, "vec index")
 
+    // Don't bother with complex dynamic indexing logic when the index is a literal and therefore static
+    if (p.isLit) {
+      return this.apply(p.litValue.intValue)
+    }
```

Comment:
@aswaterman has reminded me that we actually support out-of-bounds indices currently which means this could cause some currently working code to become an error. I'm going to change this to not accidentally change semantics and open a separate PR making those out-of-bounds indices a warning à la https://github.com/chipsalliance/chisel/pull/3033.